### 1. Write a function that inputs a number and prints the multiplication table of that number

In [171]:
def multiplication_table(n):
    for i in range(1,13):
        print('{} * {} = {}'.format(n,i,n*i))

n=int(input('Enter the number for printing the table: '))
multiplication_table(n)

Enter the number for printing the table: 10
10 * 1 = 10
10 * 2 = 20
10 * 3 = 30
10 * 4 = 40
10 * 5 = 50
10 * 6 = 60
10 * 7 = 70
10 * 8 = 80
10 * 9 = 90
10 * 10 = 100
10 * 11 = 110
10 * 12 = 120


### 2. Write a program to print twin primes less than 1000. If two consecutive odd numbers are both prime then they are known as twin primes

In [176]:
#Creating function return if a number is prime are not (True/False)
def prime(n):
    for i in range(2,n):
        if n%i==0:
            return False
    else: 
        return True

#Function to find twin prime
def twin_prime(num):  
    for i in range(2,num):
        if prime(i) and prime(i+2):
            print(i,'and',i+2)

twin_prime(1000)           

3 and 5
5 and 7
11 and 13
17 and 19
29 and 31
41 and 43
59 and 61
71 and 73
101 and 103
107 and 109
137 and 139
149 and 151
179 and 181
191 and 193
197 and 199
227 and 229
239 and 241
269 and 271
281 and 283
311 and 313
347 and 349
419 and 421
431 and 433
461 and 463
521 and 523
569 and 571
599 and 601
617 and 619
641 and 643
659 and 661
809 and 811
821 and 823
827 and 829
857 and 859
881 and 883


### 3. Write a program to find out the prime factor of a number. Example: prime factors of 56 - 2,2,2,7

In [178]:
def prime_factor(num):
    if prime(num) is True:
        return [num] 

    for i in range(2,num):
        if num%i==0:
            q=round(num/i)
            break
    t=prime_factor(q)
    t.insert(0,i)
    return t


prime_factor(1092)

[2, 2, 3, 7, 13]

### 4. Write a program to implement these formulae of permutation and combinations. Number of permutations of n objects taken r at a time: p(n,r) =n!/(n-r)!. Number of combinations of n objects taken r at a time is: c(n,r) = n!/(r!*(n-r)!)=p(n,r)/r!

In [179]:
def factorial(n):
    if n==1:
        return n
    else:
        return n*factorial(n-1)

def permutation(n,r):
    return factorial(n)/factorial(n-r)

def combination(n,r):
    return permutation(n,r)/factorial(r)

print(permutation(6,3))
print(combination(9,2))   

120.0
36.0


### 5. Write a function that converts a decimal number to binary number

https://www.vedantu.com/maths/decimal-to-binary

In [180]:
def binary_number(n):
    k=''
    while n>0:
        r=n//2
        if 2*r==n:
            k+='0'
        else:
            k+='1'
        n=r
    k=k[::-1]
    return int(k)
    
binary_number(244)

11110100

### 6. Write a function cubesum() that accepts an integer and returns the sum of the cubes of individual digits of that number. Use this function to make functions PrintArmstrong() and isArmstrong() to print Armstrong numbers and to find whether is an Armstrong number.

https://pages.mtu.edu/~shene/COURSES/cs201/NOTES/chap04/arms.html#:~:text=An%20Armstrong%20number%20of%20three,%2B%201**3%20%3D%20371.

https://www.geeksforgeeks.org/find-first-last-digits-number/

In [181]:
def cube_sum(num):
    sum_=0
    while num>0:
        last=num%10
        sum_=sum_+(last**3)
        num=num//10
        
    return sum_

#since the question is explicitly asking to use the fucntion cubesum in other function, 
#only three digit armstrong numbers  are possible

def PrintArmstrong(n):
    for i in range(100,n):
        if cube_sum(i)==i:
            print(i)
            
def isArmstrong(n):
    if cube_sum(n)==n:
        return True
    else:
        return False

PrintArmstrong(1000)
isArmstrong(371)

153
370
371
407


True

### 7. Write a function prodDigits() that inputs a number and returns the product of digits of that number 

In [182]:
def prodDigits(n):
    product=1
    while n>0:
        last = n%10
        product=product*last
        n=n//10
    return product

prodDigits(12345)

120

### 8. If all digits of a number n are multiplied by each other repeating with the product, the one digit number obtained at last is called the multicaptive digital root of n. The number of times digit need to be multiplied to reach one digit is called the multicaptive persistance of n.
Example: 86->48->32->6 (MDR 6,MPersistance 3)
         341->12->2    (MDR 2,MPersistance 2)
      

In [183]:
def product_Of_int(num):
    n=num
    p=1
    while n>0:
        last=n%10
        p*=last
        n=n//10
        
    return p


def MDR(num):
    
    n=num
    c=0
    while n>=10:
        r=product_Of_int(n)
        c+=1
        n=r
        
    print('MDR {} , MPersistance {}'.format(n,c))
    
    
MDR(86) 
MDR(341)   

MDR 6 , MPersistance 3
MDR 2 , MPersistance 2


### 9. Write a function sumPdivisor() that finds the sum of proper divisor of a number. Proper divisor of a number are those numbers by which the number is divisible, except the number itself. For example proper divisor of 36 are 1,2,3,4,6,9,18 

In [184]:
def sumPdivisor(n):
    r=0
    l=[]
    for i in range(1,n):
        if n%i==0:
            r+=i
            l.append(i)
    print(l)
    return r
            
sumPdivisor(36)    

[1, 2, 3, 4, 6, 9, 12, 18]


55

### 10. A number is called perfect if the sum of proper divisor of that number is equal to the number. For example 28 is perfect number, since 1+2+4+7+14=28. Write a program to print all the perfect numbers in given range.

In [185]:
def sumPdivisor(n):
    r=0
    for i in range(1,n):
        if n%i==0:
            r+=i
    return r

def print_perfect_num(s,e):
    for i in range(s,e+1):
        if i==sumPdivisor(i):
            print(i)
            
print_perfect_num(1,1000)

6
28
496


### 11. Two different numbers are called amicable numbers if the sum of the proper divisors of each is equal to the other number. For example 220 and 284 are amicable numbers.  Sum of proper divisors of 220 = 1+2+4+5+10+11+20+22+44+55+110 = 284 Sum of proper divisors of 284 = 1+2+4+71+142 = 220

In [186]:
def amicable_numbers(s,e):
    
    r=list()
    
    for i in range(s,e+1):
        r.append([i,sumPdivisor(i)])
        
    #print(r)    
    k=r
    l=[]
    for i in r:
        for j in k:
            if i[0]==j[1] and j[0]==i[1] and i[0]!=j[0]:
                if j[0] not in l: 
                    print(i[0],j[0])
                    l.append(i[0])
        

                    
amicable_numbers(100,10000)

    
    

220 284
1184 1210
2620 2924
5020 5564
6232 6368


### 12. Write a program which can filter odd numbers in a list by using filter function

In [187]:
def filter_odd(l):
    r= list(filter(lambda i : i%2!=0,l))
    return r

filter_odd([1,2,3,4,5,6,7,8,9])

[1, 3, 5, 7, 9]

### 13. Write a program which can map() to make a list whose elements are cube of elements in a givelist

In [188]:
def cube_of_list(l):
    r=list(map(lambda i: i**3,l))
    return r

cube_of_list([1,2,3,4,5,6,7,8,9])

[1, 8, 27, 64, 125, 216, 343, 512, 729]

### 14. Write a program which can map() and filter() to make a list whose elements are cube of even number in a given list

In [170]:
def cube_of_even(l):
    r=list(map(lambda i: i**3 ,list(filter(lambda i: i%2==0,l))))
    return r

cube_of_even([1,2,3,4,5,6,7,8,9])

[8, 64, 216, 512]